File: 2025/day6/program2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func main() {

	file, _ := os.Open("input.txt")
	scanner := bufio.NewScanner(file)

	var stringValues []string
	var operations []string
	for scanner.Scan() {
		line := scanner.Text()

		if strings.HasPrefix(line, "*") || strings.HasPrefix(line, "+") {

			var re = regexp.MustCompile(`\s+`)
			lineNormalized := strings.TrimSpace(line)
			lineNormalized = re.ReplaceAllString(lineNormalized, " ")
			operations = strings.Split(lineNormalized, " ")
			break
		}

		for i, char := range line {
			if i > len(stringValues)-1 {
				stringValues = append(stringValues, string(char))
			} else {
				stringValues[i] = fmt.Sprintf("%s%s", stringValues[i], string(char))
			}
		}
	}

	var values [][]int
	var i int
	var buffer []int
	for i < len(stringValues) {
		if strings.TrimSpace(stringValues[i]) == "" {
			values = append(values, buffer)
			buffer = []int{}
		} else {
			value, _ := strconv.Atoi(strings.TrimSpace(stringValues[i]))
			buffer = append(buffer, value)
		}
		i++
	}
	values = append(values, buffer)

	var partialAnswers []int
	var finalAnswer int
	for i, sign := range operations {
		if sign == "*" {
			answer := 1
			for _, value := range values[i] {
				answer *= value
			}
			partialAnswers = append(partialAnswers, answer)
		} else if sign == "+" {
			answer := 0
			for _, value := range values[i] {
				answer += value
			}
			partialAnswers = append(partialAnswers, answer)
		}
	}
	for _, value := range partialAnswers {
		finalAnswer += value
	}

	fmt.Println(partialAnswers)
	fmt.Println(finalAnswer)

}

File: 2025/day7/program2.py
lines = [list(line.strip()) for line in open("input.txt")]
lines = [["0" if value == "." else value for value in line] for line in lines]

for i in range(1, len(lines)):
    for j in range(len(lines[i])):
        if lines[i][j].isdigit():
            if lines[i - 1][j] == "S":
                lines[i][j] = "1"
            if lines[i - 1][j].isdigit() and int(lines[i - 1][j]) > 0:
                lines[i][j] = str(int(lines[i - 1][j]) + int(lines[i][j]))

        elif lines[i][j] == "^":
            if (
                lines[i - 1][j].isdigit()
                and int(lines[i - 1][j]) > 0
                and j > 0
                and j < len(lines[i]) - 1
            ):
                if lines[i][j - 1] != "^":
                    lines[i][j - 1] = str(int(lines[i][j - 1]) + int(lines[i - 1][j]))
                if lines[i][j + 1] != "^":
                    lines[i][j + 1] = str(int(lines[i][j + 1]) + int(lines[i - 1][j]))

counter = sum(int(value) for value in lines[-1])
print(counter)
